presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java

```diff
@@ -0,0 +1,400 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.connector.informationschema;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.Session;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.QualifiedObjectName;
+import io.prestosql.metadata.QualifiedTablePrefix;
+import io.prestosql.security.AccessControl;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.PageBuilder;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ColumnMetadata;
+import io.prestosql.spi.connector.ConnectorPageSource;
+import io.prestosql.spi.connector.ConnectorViewDefinition;
+import io.prestosql.spi.connector.SchemaTableName;
+import io.prestosql.spi.security.AccessDeniedException;
+import io.prestosql.spi.security.GrantInfo;
+import io.prestosql.spi.security.PrestoPrincipal;
+import io.prestosql.spi.security.RoleGrant;
+import io.prestosql.spi.type.Type;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Sets.union;
+import static io.prestosql.connector.informationschema.InformationSchemaMetadata.TABLES;
+import static io.prestosql.metadata.MetadataListing.listSchemas;
+import static io.prestosql.metadata.MetadataListing.listTableColumns;
+import static io.prestosql.metadata.MetadataListing.listTablePrivileges;
+import static io.prestosql.metadata.MetadataListing.listTables;
+import static io.prestosql.metadata.MetadataListing.listViews;
+import static io.prestosql.spi.security.PrincipalType.USER;
+import static io.prestosql.spi.type.TypeUtils.writeNativeValue;
+import static java.util.Objects.requireNonNull;
+
+public class InformationSchemaPageSource
+        implements ConnectorPageSource
+{
+    private enum InformationSchemaTableType
+    {
+        TABLE_COLUMNS,
+        TABLE_TABLES,
+        TABLE_VIEWS,
+        TABLE_SCHEMATA,
+        TABLE_TABLE_PRIVILEGES,
+        TABLE_ROLES,
+        TABLE_APPLICABLE_ROLES,
+        TABLE_ENABLED_ROLES
+    }
+
+    private final Session session;
+    private final Metadata metadata;
+    private final AccessControl accessControl;
+    private final String catalogName;
+    private final Optional<Iterator<QualifiedTablePrefix>> tablePrefixIterator;
+
+    private final InformationSchemaTableType tableType;
+
+    private final List<Type> columnTypes;
+    private final Queue<Page> rawPages = new ArrayDeque<>();
```

Comment:
Do you actually need a collection here? It looks like you are always creating single page at once.

Reply:
I think that's not necessarily the case. One call to `addColumnsRecordsFor` may build multiple pages.